piq/clip_iqa.py

```diff
@@ -0,0 +1,128 @@
+
+    This method is proposed to eliminate the linguistic ambiguity of the naive approach
+    (using a single prompt, e.g., "Good photo").
+
+    This method has an extension called CLIP-IQA+[2] proposed in the same research paper.
+    It uses the same approach but also fine-tunes the CLIP weights using the CoOp[3] fine-tuning algorithm.
+
+    Note:
+        The initial computation of the metric is performed in `float32` and other dtypes (i.e. `float16`, `float64`)
+        are not supported. We preserve this behaviour for reproducibility perposes. Also, at the time of writing
+        conv2d is not supported for `float16` tensors on CPU.
+
+    Warning:
+        In order to avoid implicit dtype conversion and normalization of input tensors, they are copied.
+        Note that it may consume extra memory, which might be noticible on large batch sizes.
+
+    Args:
+        data_range: Maximum value range of images (usually 1.0 or 255).
+
+    Examples:
+        >>> from piq import CLIPIQA
+        >>> clipiqa = CLIPIQA()
+        >>> x = torch.rand(1, 3, 224, 224)
+        >>> score = clipiqa(x)
+
+    References:
+        [1] Radford, Alec, et al. "Learning transferable visual models from natural language supervision."
+        International conference on machine learning. PMLR, 2021.
+        [2] Wang, Jianyi, Kelvin CK Chan, and Chen Change Loy. "Exploring CLIP for Assessing the Look
+        and Feel of Images." arXiv preprint arXiv:2207.12396 (2022).
+        [3] Zhou, Kaiyang, et al. "Learning to prompt for vision-language models." International
+        Journal of Computer Vision 130.9 (2022): 2337-2348.
+    """
+    def __init__(self, data_range: Union[float, int] = 1.) -> None:
+        super().__init__()
+
+        self.feature_extractor = clip.load().eval()
+        for param in self.feature_extractor.parameters():
+            param.requires_grad = False
+        
+        # Pre-computed tokens for prompt pairs: "Good photo.", "Bad photo.".
+        tokens = download_tensor(TOKENS_URL, os.path.expanduser("~/.cache/clip"))
+
+        anchors = self.feature_extractor.encode_text(tokens).float()
+        anchors = anchors / anchors.norm(dim=-1, keepdim=True)
+
+        self.data_range = float(data_range)
+        default_mean = torch.tensor(OPENAI_CLIP_MEAN).view(1, 3, 1, 1)
+        default_std = torch.tensor(OPENAI_CLIP_STD).view(1, 3, 1, 1)
+        self.logit_scale = self.feature_extractor.logit_scale.exp()
+
+        # Take advantage of Torch buffers. CLIPIQA.to(device) will move these to the device as well.
+        self.register_buffer("anchors", anchors)
+        self.register_buffer("default_mean", default_mean)
+        self.register_buffer("default_std", default_std)
+
+    def forward(self, x_input: torch.Tensor) -> torch.Tensor:
+        r"""Computation of CLIP-IQA metric for a given image :math:`x`.
+
+        Args:
+            x: An input tensor. Shape :math:`(N, C, H, W)` or :math:`(C, H, W)`.
```

Comment:
I would suggest to remove processing of 3D tensors to make it consistent with the rest of the metrics.

Reply:
Why would we do that it if works fine for both? I added some extra tests to verify that.